microsoft-edge/webview2/how-to/winrt-from-js.md

```diff
@@ -0,0 +1,366 @@
+
+1. Right-click the **WinRTAdapter** project, and then select **Build**.
+
+Source code is generated for namespaces or classes that you specified in the **Include filters** dialog.  That dialog populates the **Include filters** row of the **WinRTAdapter Property Pages** dialog for the **WinRTAdapter** project.
+
+
+<!-- =============================================== -->
+## Step 8. Add the host object in the webview2_sample_uwp project
+
+1. In Solution Explorer, expand the **webview2_sample_uwp** project, and then select **MainPage.xaml.cs**.
+
+1. Below the `MainPage` constructor, add the following method, to pass the WinRT object from the native side of the host app to the web side of the host app:
+
+   ```csharp
+   private async void InitializeWebView2Async()
+   {
+      await WebView2.EnsureCoreWebView2Async();
+      var dispatchAdapter = new WinRTAdapter.DispatchAdapter();
+      WebView2.CoreWebView2.AddHostObjectToScript("Windows", dispatchAdapter.WrapNamedObject("Windows", dispatchAdapter));
+   }
+   ```
+    
+1. In the `MainPage` constructor, above the `StatusUpdate("Ready");` line, add the following code:
+
+   ```csharp
+   InitializeWebView2Async();
+   ```
+
+1. Right-click the **webview2_sample_uwp** project, and then select **Set as startup project**.
+
+1. Press `Shift+Ctrl+S` to save all files.
+
+1. Press `F5` to run the sample app.  The **WebView2 WinUI 2 UWP Sample** window opens.
+
+   ![The WebView2 WinUI 2 UWP Sample window.](winrt-from-js-images/webview2-winui-2-uwp-sample-app-window.png)
+
+The host app's web-side code, including DevTools **Console**, can now call methods and properties of the specified namespaces or classes of the host object.
+
+
+<!-- =============================================== -->
+## Step 9. Call methods and properties on the host object from web-side JavaScript
+
+Next, use the DevTools **Console** to demonstrate that web-side code can call the included, specified host-side API items.
+
+1. Click in the main part of the WebView2 sample app window to give it focus, and then press `Ctrl+Shift+I` to open Microsoft Edge DevTools.  Or, right-click the page, and then select **Inspect**.  The Microsoft Edge DevTools window opens.
+
+1. If the Microsoft Edge DevTools window isn't visible, press `Alt+Tab` to display the DevTools window.  If needed, move the DevTools window.
+
+1. In the **Console** of DevTools, paste the following code, and then press `Enter`:
+
+   ```javascript
+   const Windows = chrome.webview.hostObjects.sync.Windows;
+   (new Windows.Globalization.Language("en-US")).displayName;
+   ```
+
+   The **Console** outputs a language name string, such as `English (United States)`, demonstrating that your app's host-side code can be called from web-side code.
+
+   ![Using the DevTools Console to test calling native-side code from web-side code.](winrt-from-js-images/devtools-console-calling-native-side-code.png)
+
+1. Close the UWP sample app.  The DevTools window also closes.
+
```

Comment:
We're missing the section of the wiki where we said 

# Make AddHostObjectToScript act more like Chakra WinRT projection

The AddHostObjectToScript feature defaults to using async and verbose proxies (you can read more about that in the [AddHostObjectToScript documentation](https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2?view=webview2-1.0.1072.54#addhostobjecttoscript) ). To make it act more like the Chakra WinRT projection you'll need to set a few properties:

* `chrome.webview.hostObjects.option.defaultSyncProxy` Proxies may either be async or sync. Normally we know when calling a method on a sync proxy that the result should also be a sync proxy. But in some cases we lose that context, like providing a reference to a function to native code and native code later calling that function. In these cases the proxy will be async unless this property is set.
 * `chrome.webview.hostObjects.options.forceAsyncMethodMatches` This is an array of regular expressions. If you call a method on a sync proxy, the method call will actually be performed asynchronously if the method name matches a string or regular expression in this array. Setting this to [/Async$/] will make any method ending with Async be an asynchronous method call. If an async method doesn't match here and isn't forced to be async, it will be invoked synchronously, blocking execution of the calling JavaScript and then returning the resolution of the promise rather than a promise.
 * `chrome.webview.hostObjects.options.ignoreMemberNotFoundError` If you attempt to get the value of a property of proxy and the property does not exist on the corresponding native class, you will get an exception. Unless you set this property to true in which case it will match Chakra WinRT projection behavior (and general JavaScript behavior) and return undefined with no error.

Chakra WinRT projection put the WinRT namespaces directly on the root object. AddHostObjectToScript places async root proxies on chrome.webview.hostObjects and sync root proxies on chrome.webview.hostObjects.sync. To access them where Chakra WinRT projection code would expect, you can assign them over to the root object. For example:

`window.Windows = chrome.webview.hostObjects.sync.Windows;`

To ensure the JavaScript that sets all this up is executed before anything else, you can either add it to your JavaScript, or you can tell WebView2 to inject it for you before running any other script using the `CoreWebView2.AddScriptToExecuteOnDocumentCreatedAsync` method.

An all of the above example:

```
webView->CoreWebView2->AddScriptToExecuteOnDocumentCreatedAsync(
            LR"(
                (() => {
                    if (chrome && chrome.webview) {
                        console.log('Setting up WinRT projection options');
                        chrome.webview.hostObjects.options.defaultSyncProxy = true;
                        chrome.webview.hostObjects.options.forceAsyncMethodMatches = [/Async$/,/AsyncWithSpeller$/];
                        chrome.webview.hostObjects.options.ignoreMemberNotFoundError = true;

                        window.CortanaApp = chrome.webview.hostObjects.sync.CortanaApp;
                        window.Cortana = chrome.webview.hostObjects.sync.Cortana;
                        window.Windows = chrome.webview.hostObjects.sync.Windows;
                    }
                })();
            )"));
```